File: repowire/daemon/state/operations.py
"""SQLite-backed durable operation lifecycle records."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any
from uuid import uuid4

from repowire.daemon.state.database import StateDatabase
from repowire.daemon.work_store import json_dumps, json_loads, now_iso


def new_operation_id() -> str:
    return f"op-{uuid4().hex[:12]}"


@dataclass(frozen=True)
class Operation:
    operation_id: str
    kind: str
    state: str
    target: dict[str, Any] = field(default_factory=dict)
    strategy: str | None = None
    attempts: list[dict[str, Any]] = field(default_factory=list)
    result: dict[str, Any] = field(default_factory=dict)
    error: dict[str, Any] = field(default_factory=dict)
    provenance: dict[str, Any] = field(default_factory=dict)
    created_at: str = ""
    updated_at: str = ""
    completed_at: str | None = None

    def status(self) -> dict[str, Any]:
        return {
            "operation_id": self.operation_id,
            "kind": self.kind,
            "state": self.state,
            "target": self.target,
            "strategy": self.strategy,
            "attempts": self.attempts,
            "result": self.result,
            "error": self.error,
            "provenance": self.provenance,
            "created_at": self.created_at,
            "updated_at": self.updated_at,
            "completed_at": self.completed_at,
        }


class SQLiteOperationStore:
    """Repository for internal command/operation audit state."""

    def __init__(self, db: StateDatabase) -> None:
        self._conn = db.conn

    @staticmethod
    def _row_to_operation(row) -> Operation | None:
        if row is None:
            return None
        return Operation(
            operation_id=row["operation_id"],
            kind=row["kind"],
            state=row["state"],
            target=json_loads(row["target_json"], {}),
            strategy=row["strategy"],
            attempts=json_loads(row["attempts_json"], []),
            result=json_loads(row["result_json"], {}),
            error=json_loads(row["error_json"], {}),
            provenance=json_loads(row["provenance_json"], {}),
            created_at=row["created_at"],
            updated_at=row["updated_at"],
            completed_at=row["completed_at"],
        )

    def create(
        self,
        *,
        kind: str,
        target: dict[str, Any] | None = None,
        provenance: dict[str, Any] | None = None,
    ) -> Operation:
        now = now_iso()
        operation = Operation(
            operation_id=new_operation_id(),
            kind=kind,
            state="queued",
            target=target or {},
            provenance=provenance or {},
            created_at=now,
            updated_at=now,
        )
        with self._conn:
            self._conn.execute(
                """
                INSERT INTO operations(
                    operation_id, kind, state, target_json, strategy,
                    attempts_json, result_json, error_json, provenance_json,
                    created_at, updated_at, completed_at
                ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
                """,
                (
                    operation.operation_id,
                    operation.kind,
                    operation.state,
                    json_dumps(operation.target),
                    operation.strategy,
                    json_dumps(operation.attempts),
                    json_dumps(operation.result),
                    json_dumps(operation.error),
                    json_dumps(operation.provenance),
                    operation.created_at,
                    operation.updated_at,
                    operation.completed_at,
                ),
            )
        return operation

    def get(self, operation_id: str) -> Operation | None:
        row = self._conn.execute(
            "SELECT * FROM operations WHERE operation_id = ?",
            (operation_id,),
        ).fetchone()
        return self._row_to_operation(row)

    def list_all(self, *, kind: str | None = None, state: str | None = None) -> list[Operation]:
        clauses: list[str] = []
        params: list[str] = []
        if kind is not None:
            clauses.append("kind = ?")
            params.append(kind)
        if state is not None:
            clauses.append("state = ?")
            params.append(state)
        where = f"WHERE {' AND '.join(clauses)}" if clauses else ""
        rows = self._conn.execute(
            f"SELECT * FROM operations {where} ORDER BY updated_at DESC",
            params,
        ).fetchall()
        return [
            operation
            for row in rows
            if (operation := self._row_to_operation(row)) is not None
        ]

    def start_attempt(
        self,
        operation_id: str,
        *,
        strategy: str | None = None,
        detail: dict[str, Any] | None = None,
    ) -> Operation | None:
        operation = self.get(operation_id)
        if operation is None:
            return None
        attempts = list(operation.attempts)
        attempts.append(
            {
                "attempt_id": f"op-attempt-{uuid4().hex[:12]}",
                "state": "running",
                "strategy": strategy,
                "detail": detail or {},
                "started_at": now_iso(),
                "completed_at": None,
                "error": {},
            }
        )
        return self._update(
            operation,
            state="running",
            strategy=strategy if strategy is not None else operation.strategy,
            attempts=attempts,
        )

    def complete(
        self,
        operation_id: str,
        *,
        strategy: str | None = None,
        result: dict[str, Any] | None = None,
    ) -> Operation | None:
        operation = self.get(operation_id)
        if operation is None:
            return None
        now = now_iso()
        attempts = list(operation.attempts)
        if attempts:
            attempts[-1] = {
                **attempts[-1],
                "state": "completed",
                "completed_at": now,
                "result": result or {},
            }
        return self._update(
            operation,
            state="completed",
            strategy=strategy if strategy is not None else operation.strategy,
            attempts=attempts,
            result=result or {},
            completed_at=now,
        )

    def fail(
        self,
        operation_id: str,
        *,
        state: str = "failed",
        strategy: str | None = None,
        error: dict[str, Any] | None = None,
    ) -> Operation | None:
        operation = self.get(operation_id)
        if operation is None:
            return None
        now = now_iso()
        attempts = list(operation.attempts)
        if attempts:
            attempts[-1] = {
                **attempts[-1],
                "state": state,
                "completed_at": now,
                "error": error or {},
            }
        return self._update(
            operation,
            state=state,
            strategy=strategy if strategy is not None else operation.strategy,
            attempts=attempts,
            error=error or {},
            completed_at=now,
        )

    def _update(
        self,
        operation: Operation,
        *,
        state: str,
        strategy: str | None,
        attempts: list[dict[str, Any]] | None = None,
        result: dict[str, Any] | None = None,
        error: dict[str, Any] | None = None,
        completed_at: str | None = None,
    ) -> Operation | None:
        now = now_iso()
        with self._conn:
            self._conn.execute(
                """
                UPDATE operations
                SET state = ?, strategy = ?, attempts_json = ?, result_json = ?,
                    error_json = ?, updated_at = ?, completed_at = ?
                WHERE operation_id = ?
                """,
                (
                    state,
                    strategy,
                    json_dumps(attempts if attempts is not None else operation.attempts),
                    json_dumps(result if result is not None else operation.result),
                    json_dumps(error if error is not None else operation.error),
                    now,
                    completed_at,
                    operation.operation_id,
                ),
            )
        return self.get(operation.operation_id)

File: daemon-go/state/operations.go
package state

import (
	"context"
	"database/sql"
	"encoding/json"
	"fmt"
	"maps"
)

// opJSONLoadsArray mirrors json_loads(raw, []) — returns a slice, [] on empty/non-array.
func opJSONLoadsArray(raw string) []map[string]any {
	if raw == "" {
		return []map[string]any{}
	}
	var v any
	if err := json.Unmarshal([]byte(raw), &v); err != nil {
		return []map[string]any{}
	}
	arr, ok := v.([]any)
	if !ok {
		return []map[string]any{}
	}
	out := make([]map[string]any, 0, len(arr))
	for _, item := range arr {
		if m, ok := item.(map[string]any); ok {
			out = append(out, m)
		} else {
			out = append(out, map[string]any{})
		}
	}
	return out
}

// Operation mirrors operations.Operation. JSON columns are decoded into
// generic maps/slices since their shapes are caller-defined.
type Operation struct {
	OperationID string
	Kind        string
	State       string
	Target      map[string]any
	Strategy    *string
	Attempts    []map[string]any
	Result      map[string]any
	Error       map[string]any
	Provenance  map[string]any
	CreatedAt   string
	UpdatedAt   string
	CompletedAt *string
}

func scanOperation(row interface {
	Scan(dest ...any) error
}) (*Operation, error) {
	var (
		operationID  string
		kind         string
		state        string
		targetJSON   string
		strategy     sql.NullString
		attemptsJSON string
		resultJSON   string
		errorJSON    string
		provJSON     string
		createdAt    string
		updatedAt    string
		completedAt  sql.NullString
	)
	if err := row.Scan(
		&operationID, &kind, &state, &targetJSON, &strategy,
		&attemptsJSON, &resultJSON, &errorJSON, &provJSON,
		&createdAt, &updatedAt, &completedAt,
	); err != nil {
		return nil, err
	}
	op := &Operation{
		OperationID: operationID,
		Kind:        kind,
		State:       state,
		Target:      decodeJSONObject(targetJSON),
		Attempts:    opJSONLoadsArray(attemptsJSON),
		Result:      decodeJSONObject(resultJSON),
		Error:       decodeJSONObject(errorJSON),
		Provenance:  decodeJSONObject(provJSON),
		CreatedAt:   createdAt,
		UpdatedAt:   updatedAt,
	}
	if strategy.Valid {
		s := strategy.String
		op.Strategy = &s
	}
	if completedAt.Valid {
		c := completedAt.String
		op.CompletedAt = &c
	}
	return op, nil
}

const operationColumns = `operation_id, kind, state, target_json, strategy,
	attempts_json, result_json, error_json, provenance_json,
	created_at, updated_at, completed_at`

// CreateOperation mirrors SQLiteOperationStore.create: inserts a queued operation
// with a fresh op id and now-stamped created_at/updated_at.
func (s *Store) CreateOperation(ctx context.Context, kind string, target, provenance map[string]any) (*Operation, error) {
	if target == nil {
		target = map[string]any{}
	}
	if provenance == nil {
		provenance = map[string]any{}
	}
	now := nowISO()
	op := &Operation{
		OperationID: newID("op-"),
		Kind:        kind,
		State:       "queued",
		Target:      target,
		Attempts:    []map[string]any{},
		Result:      map[string]any{},
		Error:       map[string]any{},
		Provenance:  provenance,
		CreatedAt:   now,
		UpdatedAt:   now,
	}

	targetJSON, err := marshalJSON(op.Target)
	if err != nil {
		return nil, fmt.Errorf("marshal target: %w", err)
	}
	attemptsJSON, err := marshalJSON(op.Attempts)
	if err != nil {
		return nil, fmt.Errorf("marshal attempts: %w", err)
	}
	resultJSON, err := marshalJSON(op.Result)
	if err != nil {
		return nil, fmt.Errorf("marshal result: %w", err)
	}
	errorJSON, err := marshalJSON(op.Error)
	if err != nil {
		return nil, fmt.Errorf("marshal error: %w", err)
	}
	provJSON, err := marshalJSON(op.Provenance)
	if err != nil {
		return nil, fmt.Errorf("marshal provenance: %w", err)
	}

	const q = `INSERT INTO operations(
		operation_id, kind, state, target_json, strategy,
		attempts_json, result_json, error_json, provenance_json,
		created_at, updated_at, completed_at
	) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`
	_, err = s.db.ExecContext(ctx, q,
		op.OperationID, op.Kind, op.State, targetJSON, nil,
		attemptsJSON, resultJSON, errorJSON, provJSON,
		op.CreatedAt, op.UpdatedAt, nil,
	)
	if err != nil {
		return nil, fmt.Errorf("insert operation %s: %w", op.OperationID, err)
	}
	return op, nil
}

// GetOperation mirrors SQLiteOperationStore.get: returns (nil, nil) when missing.
func (s *Store) GetOperation(ctx context.Context, operationID string) (*Operation, error) {
	const q = `SELECT ` + operationColumns + ` FROM operations WHERE operation_id = ?`
	op, err := scanOperation(s.db.QueryRowContext(ctx, q, operationID))
	if err != nil {
		if err == sql.ErrNoRows {
			return nil, nil
		}
		return nil, fmt.Errorf("get operation %s: %w", operationID, err)
	}
	return op, nil
}

// ListOperations mirrors SQLiteOperationStore.list_all: optional kind/state filters
// (empty string means no filter), ordered by updated_at DESC.
func (s *Store) ListOperations(ctx context.Context, kind, state string) ([]*Operation, error) {
	q := `SELECT ` + operationColumns + ` FROM operations`
	var clauses []string
	var args []any
	if kind != "" {
		clauses = append(clauses, "kind = ?")
		args = append(args, kind)
	}
	if state != "" {
		clauses = append(clauses, "state = ?")
		args = append(args, state)
	}
	if len(clauses) > 0 {
		q += " WHERE " + clauses[0]
		for _, c := range clauses[1:] {
			q += " AND " + c
		}
	}
	q += " ORDER BY updated_at DESC"

	rows, err := s.db.QueryContext(ctx, q, args...)
	if err != nil {
		return nil, fmt.Errorf("list operations: %w", err)
	}
	defer rows.Close()

	var out []*Operation
	for rows.Next() {
		op, err := scanOperation(rows)
		if err != nil {
			return nil, fmt.Errorf("scan operation: %w", err)
		}
		out = append(out, op)
	}
	if err := rows.Err(); err != nil {
		return nil, fmt.Errorf("iterate operations: %w", err)
	}
	return out, nil
}

// StartAttempt mirrors SQLiteOperationStore.start_attempt: appends a running
// attempt and transitions the operation to running. Returns (nil, nil) when missing.
func (s *Store) StartAttempt(ctx context.Context, operationID string, strategy *string, detail map[string]any) (*Operation, error) {
	op, err := s.GetOperation(ctx, operationID)
	if err != nil || op == nil {
		return op, err
	}
	if detail == nil {
		detail = map[string]any{}
	}
	attempts := append([]map[string]any{}, op.Attempts...)
	attempts = append(attempts, map[string]any{
		"attempt_id":   newID("op-attempt-"),
		"state":        "running",
		"strategy":     opStrategyValue(strategy),
		"detail":       detail,
		"started_at":   nowISO(),
		"completed_at": nil,
		"error":        map[string]any{},
	})
	nextStrategy := op.Strategy
	if strategy != nil {
		nextStrategy = strategy
	}
	return s.updateOperation(ctx, op, "running", nextStrategy, attempts, nil, nil, nil)
}

// CompleteOperation mirrors SQLiteOperationStore.complete: marks the last attempt
// and the operation completed. Returns (nil, nil) when missing.
func (s *Store) CompleteOperation(ctx context.Context, operationID string, strategy *string, result map[string]any) (*Operation, error) {
	op, err := s.GetOperation(ctx, operationID)
	if err != nil || op == nil {
		return op, err
	}
	if result == nil {
		result = map[string]any{}
	}
	now := nowISO()
	attempts := append([]map[string]any{}, op.Attempts...)
	if n := len(attempts); n > 0 {
		last := opCloneMap(attempts[n-1])
		last["state"] = "completed"
		last["completed_at"] = now
		last["result"] = result
		attempts[n-1] = last
	}
	nextStrategy := op.Strategy
	if strategy != nil {
		nextStrategy = strategy
	}
	return s.updateOperation(ctx, op, "completed", nextStrategy, attempts, result, nil, &now)
}

// FailOperation mirrors SQLiteOperationStore.fail: marks the last attempt and the
// operation with the given state (empty defaults to "failed"). Returns (nil, nil) when missing.
func (s *Store) FailOperation(ctx context.Context, operationID, state string, strategy *string, opErr map[string]any) (*Operation, error) {
	op, err := s.GetOperation(ctx, operationID)
	if err != nil || op == nil {
		return op, err
	}
	if state == "" {
		state = "failed"
	}
	if opErr == nil {
		opErr = map[string]any{}
	}
	now := nowISO()
	attempts := append([]map[string]any{}, op.Attempts...)
	if n := len(attempts); n > 0 {
		last := opCloneMap(attempts[n-1])
		last["state"] = state
		last["completed_at"] = now
		last["error"] = opErr
		attempts[n-1] = last
	}
	nextStrategy := op.Strategy
	if strategy != nil {
		nextStrategy = strategy
	}
	return s.updateOperation(ctx, op, state, nextStrategy, attempts, nil, opErr, &now)
}

// updateOperation mirrors SQLiteOperationStore._update: persists the mutated
// columns with a fresh updated_at and re-reads the row.
func (s *Store) updateOperation(
	ctx context.Context,
	op *Operation,
	state string,
	strategy *string,
	attempts []map[string]any,
	result map[string]any,
	opErr map[string]any,
	completedAt *string,
) (*Operation, error) {
	now := nowISO()
	if attempts == nil {
		attempts = op.Attempts
	}
	if result == nil {
		result = op.Result
	}
	if opErr == nil {
		opErr = op.Error
	}

	attemptsJSON, err := marshalJSON(attempts)
	if err != nil {
		return nil, fmt.Errorf("marshal attempts: %w", err)
	}
	resultJSON, err := marshalJSON(result)
	if err != nil {
		return nil, fmt.Errorf("marshal result: %w", err)
	}
	errorJSON, err := marshalJSON(opErr)
	if err != nil {
		return nil, fmt.Errorf("marshal error: %w", err)
	}

	const q = `UPDATE operations
		SET state = ?, strategy = ?, attempts_json = ?, result_json = ?,
		    error_json = ?, updated_at = ?, completed_at = ?
		WHERE operation_id = ?`
	_, err = s.db.ExecContext(ctx, q,
		state, strOrNil(strategy), attemptsJSON, resultJSON,
		errorJSON, now, strOrNil(completedAt),
		op.OperationID,
	)
	if err != nil {
		return nil, fmt.Errorf("update operation %s: %w", op.OperationID, err)
	}
	return s.GetOperation(ctx, op.OperationID)
}

// opStrategyValue renders an optional strategy as a JSON-friendly value (nil → null),
// matching Python's None passthrough into attempt records.
func opStrategyValue(s *string) any {
	if s == nil {
		return nil
	}
	return *s
}

// opCloneMap shallow-copies a map so attempt mutation doesn't alias the decoded slice.
func opCloneMap(m map[string]any) map[string]any {
	return maps.Clone(m)
}
